packages/abi-coder/src/types/JsonAbiNew.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Types for Fuel JSON ABI Format as defined on:
+ * https://github.com/FuelLabs/fuel-specs/blob/master/src/abi/json-abi-format.md
+ */
+export interface JsonAbi {
+  readonly specVersion: string;
+  readonly encodingVersion: string;
+  readonly programType: string;
+  readonly concreteTypes: readonly ConcreteType[];
+  readonly metadataTypes: readonly MetadataType[];
+  readonly functions: readonly AbiFunction[];
+  readonly loggedTypes: readonly LoggedType[];
+  readonly messagesTypes: readonly MessageType[];
+  readonly configurables: readonly Configurable[];
+}
+
+export interface ConcreteType {
+  readonly type: string;
+  readonly concreteTypeId: string;
+  readonly metadataTypeId?: number;
```

Comment:
Per the [spec](https://github.com/FuelLabs/fuels-ts/pull/2856/files#diff-91cdfcdf704955525b9cd0988d8abef2b29fb92406e961e786d571b0779786f5R20), this should be a string (hash). It would help reduce complexity if this were consistent.

Reply:
The link you gave us points to this same field the comment is on...

Regardless, the [spec](https://github.com/FuelLabs/fuel-specs/blob/master/src/abi/json-abi-format.md) says this for `concreteType.metadataTypeId`:
> "metadataTypeId": the type metadata declaration ID of this type, if the type metadata has components or is generic, otherwise nonexistent.

And it defines `metadataType.metadataTypeId` as this:
> "metadataTypeId": a unique integer ID.

So this conforms with the specification.

Reply:
Sorry, updated the link. It also says this:

```
- _type metadata declaration_: the declaration or definition of a type which can be generic. `struct Foo { .. }` and `struct Bar<T> { .. }` in the example above are both type declarations. The metadata declaration contains component details about the type. And a single _type metadata declaration_ is generated per type, even for generic types.
```

Which makes it sound like "type metadata declaration ID" is the string hash, not an array index.

Reply:
The string hashes are exclusively related to concrete types (hashes of their `type` property). The spec doesn't mention array indexes in relation to `metadataTypeId`s, it just defines them as a field of the object. I wouldn't rely on that correlation to build anything as it might change: a metadata type with id `12` might be put onto index `24` and it would still be spec-compliant.